ntoskrnl/io/debug.c

```diff
@@ -0,0 +1,561 @@
+            break;
+
+        case CmResourceTypeBusNumber:
+            DPRINT1("[%p:%X:%X] BUS: Min %X Max %X Len %X\n", Descriptor, Descriptor->Option, Descriptor->ShareDisposition, Descriptor->u.BusNumber.MinBusNumber, Descriptor->u.BusNumber.MaxBusNumber, Descriptor->u.BusNumber.Length);
+            break;
+
+        case CmResourceTypeConfigData:
+            DPRINT1("[%p:%X:%X] CFG: Priority %X\n", Descriptor, Descriptor->Option, Descriptor->ShareDisposition, Descriptor->u.ConfigData.Priority);
+            break;
+
+        case CmResourceTypeDevicePrivate:
+            DPRINT1("[%p:%X:%X] DAT: %X %X %X\n", Descriptor, Descriptor->Option, Descriptor->ShareDisposition, Descriptor->u.DevicePrivate.Data[0], Descriptor->u.DevicePrivate.Data[1], Descriptor->u.DevicePrivate.Data[2]);
+            break;
+
+        default:
+            DPRINT1("%s[%p:%X:%X]. Unknown type %X\n", Descriptor, Descriptor->Option, Descriptor->ShareDisposition, Descriptor->Type);
+            break;
+    }
+}
+
+/* PipDumpResourceRequirementsList() displays information about a Io List 
+
+   DebugLevel: 0 - always dump
+               1 - dump if not defined NDEBUG
+*/
+VOID
+NTAPI
+PipDumpResourceRequirementsList(
+    _In_ PIO_RESOURCE_REQUIREMENTS_LIST IoResource,
+    _In_ ULONG DebugLevel)
+{
+    PIO_RESOURCE_LIST AltList;
+    PIO_RESOURCE_DESCRIPTOR Descriptor;
+    ULONG ix;
+    ULONG jx;
+
+    PAGED_CODE();
+
+    if (DebugLevel != 0)
+    {
+#ifdef NDEBUG
+        return;
+#endif
+    }
+
+    if (IoResource == NULL)
+    {
+        DPRINT1("PipDumpResourceRequirementsList: IoResource == 0\n");
+        return;
+    }
+
+    DPRINT1("Dump RequirementsList: IoResource %p\n", IoResource);
+    DPRINT1("Interface %X Bus %X Slot %X AlternativeLists %X\n",
+            IoResource->InterfaceType,
+            IoResource->BusNumber,
+            IoResource->SlotNumber,
+            IoResource->AlternativeLists);
+
+    AltList = &IoResource->List[0];
+
+    if (IoResource->AlternativeLists < 1)
+    {
+        DPRINT1("PipDumpResourceRequirementsList: AlternativeLists < 1\n");
+        return;
+    }
+
+    for (ix = 0; ix < IoResource->AlternativeLists; ix++)
+    {
+        DPRINT1("AltList %p, AltList->Count %X\n", AltList, AltList->Count);
+
+        for (jx = 0; jx < AltList->Count; jx++)
+        {
+            Descriptor = &AltList->Descriptors[jx];
+            PipDumpIoResourceDescriptor(Descriptor, DebugLevel);
+        }
+
+        AltList = (PIO_RESOURCE_LIST)(AltList->Descriptors + AltList->Count);
+        DPRINT1("End Descriptors %p\n", AltList);
+    }
+}
+
+/* Gets the name of the device node state. */
+
+PWSTR
+NTAPI
+PipGetDeviceNodeStateName(
+    _In_ PNP_DEVNODE_STATE State)
+{
+    PWSTR StateName;
+
+    switch (State)
+    {
+        case DeviceNodeUnspecified:
+            StateName = L"DeviceNodeUnspecified";
+            break;
```

Comment:
I think you can directly do a `return L"whatever";` for each case.

Reply:
Ok.